Throw on table splitting with multiple roots

When using table splitting with more than one child entity it does not work.
For example, if I have an parent entity, with two children that are configured in table splitting, it will work if I add the parent alone or the parent with only one of the children, but if I set the two children, it will throw  a `InvalidOperationException` (exception details at the end).

In the scenario below marked as [ERROR], I notice how right after calling `ctx.Set<PurchaseOrder>().Add(po);`, the property `po.Cancellation` has become `null`, which is very odd.

Database schema was correctly generated.

### Steps to reproduce
#### Model
```c#
public class PurchaseOrder
{
    public int Id { get; set; }
    public string Status { get; set; }
    public CancellationTicket Cancellation { get; set; }
    public Customer Customer { get; set; }
}

public class CancellationTicket
{
    public string CancelReason { get; set; }
    public DateTime? CancelDate { get; set; }
}

public class Customer
{
    public string Name { get; set; }
    public string Surname { get; set; }
}
```
#### Mapping
```c#
class PurchaseOrderConfig : IEntityTypeConfiguration<PurchaseOrder>
{
    public void Configure(EntityTypeBuilder<PurchaseOrder> builder)
    {
        builder.ToTable("purchase_order");
        builder.HasKey(x => x.Id);
        builder.Property(x => x.Id).HasColumnName("id").ForSqlServerUseSequenceHiLo();
        builder.Property(x => x.Status).HasColumnName("status");
        builder.HasOne(x => x.Cancellation).WithOne().HasForeignKey<PurchaseOrder>(x => x.Id);
        builder.HasOne(x => x.Customer).WithOne().HasForeignKey<PurchaseOrder>(x => x.Id);
    }
}

class PurchaseOrderCancellationConfig : IEntityTypeConfiguration<CancellationTicket>
{
    public void Configure(EntityTypeBuilder<CancellationTicket> builder)
    {
        builder.ToTable("purchase_order");
        builder.Property(x => x.CancelReason).HasColumnName("cancel_reason").HasMaxLength(500);
        builder.Property(x => x.CancelDate).HasColumnName("cancel_date").HasColumnType("datetime2");

        builder.Property<int>("Id").HasColumnName("id").ForSqlServerUseSequenceHiLo();
        builder.HasKey("Id");
    }
}

class CustomerConfig : IEntityTypeConfiguration<Customer>
{
    public void Configure(EntityTypeBuilder<Customer> builder)
    {
        builder.ToTable("purchase_order");
        builder.Property(x => x.Name).HasColumnName("customer_name").HasMaxLength(100);
        builder.Property(x => x.Surname).HasColumnName("customer_surname").HasMaxLength(100);

        builder.Property<int>("Id").HasColumnName("id").ForSqlServerUseSequenceHiLo();
        builder.HasKey("Id");
    }
}
```
#### Generated schema
```sql
CREATE TABLE [dbo].[purchase_order](
	[id] [int] NOT NULL,
	[cancel_date] [datetime2](7) NULL,
	[cancel_reason] [nvarchar](500) NULL,
	[customer_name] [nvarchar](100) NULL,
	[customer_surname] [nvarchar](100) NULL,
	[status] [nvarchar](max) NULL,
 CONSTRAINT [PK_purchase_order] PRIMARY KEY CLUSTERED 
(
	[id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]
GO
```
### Scenarios
#### [OK] Add main entity with only first child
```c#
var po = new PurchaseOrder()
{
    Status = "hello",
    Cancellation = new CancellationTicket()
    {
        CancelDate = DateTime.Now,
        CancelReason = "reason"
    }
};
ctx.Set<PurchaseOrder>().Add(po);
ctx.SaveChanges();
```
#### [OK] Add main entity with only second child
```c#
var po = new PurchaseOrder()
{
    Status = "hello",
    Customer = new Customer()
    {
        Name = "name2",
        Surname ="surname2"
    }
};
ctx.Set<PurchaseOrder>().Add(po);
ctx.SaveChanges();
```

#### [ERROR] Add main entity with both children
```c#
var po = new PurchaseOrder()
{
    Status = "hello3",
    Customer = new Customer()
    {
        Name = "name3",
        Surname ="surname3"
    },
    Cancellation = new CancellationTicket()
    {
        CancelDate = DateTime.Now,
        CancelReason = "reason 3"
    }
};
ctx.Set<PurchaseOrder>().Add(po);
ctx.SaveChanges();
```
ERROR:
```
System.InvalidOperationException occurred
  HResult=0x80131509
  Message=The entity of 'CancellationTicket' is sharing the table 'purchase_order' with 'PurchaseOrder', but there is no entity of this type with the same key value that has been marked as 'Added'. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the key values.
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at Microsoft.EntityFrameworkCore.Update.Internal.ModificationCommandIdentityMap.Validate(Boolean sensitiveLoggingEnabled)
   at Microsoft.EntityFrameworkCore.Update.Internal.CommandBatchPreparer.CreateModificationCommands(IReadOnlyList`1 entries, Func`1 generateParameterName)
   at Microsoft.EntityFrameworkCore.Update.Internal.CommandBatchPreparer.<BatchCommands>d__8.MoveNext()
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(Tuple`2 parameters)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState,TResult](IExecutionStrategy strategy, TState state, Func`2 operation)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabase.SaveChanges(IReadOnlyList`1 entries)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges()
   at EFCoreBug.Program.Main(String[] args) in C:\Users\valeriano.tortola\documents\visual studio 2017\Projects\EFCoreBug\Program.cs:line 51
```

### Further technical details
EF Core version: 2.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: VS 2017 15.3

--- a/src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs
+++ b/src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs
@@ -190,20 +190,42 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
                 return;
             }
 
-            var firstValidatedType = mappedTypes[0];
+            var unvalidatedTypes = new HashSet<IEntityType>(mappedTypes);
+            IEntityType root = null;
+            foreach (var mappedType in mappedTypes)
+            {
+                if (mappedType.BaseType != null
+                    || mappedType.FindForeignKeys(mappedType.FindPrimaryKey().Properties)
+                        .Any(fk => fk.PrincipalKey.IsPrimaryKey()
+                                   && fk.PrincipalEntityType != mappedType
+                                   && unvalidatedTypes.Contains(fk.PrincipalEntityType)))
+                {
+                    continue;
+                }
+
+                if (root != null)
+                {
+                    throw new InvalidOperationException(
+                        RelationalStrings.IncompatibleTableNoRelationship(
+                            tableName,
+                            mappedType.DisplayName(),
+                            root.DisplayName()));
+                }
+                root = mappedType;
+            }
+
+            unvalidatedTypes.Remove(root);
             var typesToValidate = new Queue<IEntityType>();
-            typesToValidate.Enqueue(firstValidatedType);
-            var unvalidatedTypes = new HashSet<IEntityType>(mappedTypes.Skip(1));
+            typesToValidate.Enqueue(root);
+
             while (typesToValidate.Count > 0)
             {
                 var entityType = typesToValidate.Dequeue();
                 var typesToValidateLeft = typesToValidate.Count;
-                var nextTypeSet = unvalidatedTypes.Where(
-                    unvalidatedType =>
-                        entityType.RootType() == unvalidatedType.RootType()
-                        || IsIdentifyingPrincipal(entityType, unvalidatedType)
-                        || IsIdentifyingPrincipal(unvalidatedType, entityType));
-                foreach (var nextEntityType in nextTypeSet)
+                var directlyConnectedTypes = unvalidatedTypes.Where(unvalidatedType =>
+                    entityType.IsAssignableFrom(unvalidatedType)
+                    || IsIdentifyingPrincipal(unvalidatedType, entityType));
+                foreach (var nextEntityType in directlyConnectedTypes)
                 {
                     var key = entityType.FindPrimaryKey();
                     var otherKey = nextEntityType.FindPrimaryKey();
@@ -240,41 +262,31 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
                     RelationalStrings.IncompatibleTableNoRelationship(
                         tableName,
                         invalidEntityType.DisplayName(),
-                        firstValidatedType.DisplayName(),
-                        Property.Format(invalidEntityType.FindPrimaryKey().Properties),
-                        Property.Format(firstValidatedType.FindPrimaryKey().Properties)));
+                        root.DisplayName()));
             }
         }
 
         private static bool IsIdentifyingPrincipal(IEntityType dependEntityType, IEntityType principalEntityType)
         {
-            var identifyingForeignKeys = new Queue<IForeignKey>(
-                dependEntityType.FindForeignKeys(dependEntityType.FindPrimaryKey().Properties));
-            while (identifyingForeignKeys.Count > 0)
+            foreach (var foreignKey in dependEntityType.FindForeignKeys(dependEntityType.FindPrimaryKey().Properties))
             {
-                var fk = identifyingForeignKeys.Dequeue();
-                if (fk.PrincipalKey.IsPrimaryKey())
+                if (!foreignKey.PrincipalKey.IsPrimaryKey()
+                    || foreignKey.PrincipalEntityType != principalEntityType)
                 {
-                    if (fk.PrincipalEntityType == principalEntityType)
-                    {
-                        if (principalEntityType.BaseType != null)
-                        {
-                            // #8973
-                            throw new InvalidOperationException(
-                                RelationalStrings.IncompatibleTableDerivedPrincipal(
-                                    dependEntityType.Relational().TableName,
-                                    dependEntityType.DisplayName(),
-                                    principalEntityType.DisplayName(),
-                                    principalEntityType.RootType().DisplayName()));
-                        }
-                        return true;
-                    }
-
-                    foreach (var principalFk in fk.PrincipalEntityType.FindForeignKeys(fk.PrincipalEntityType.FindPrimaryKey().Properties))
-                    {
-                        identifyingForeignKeys.Enqueue(principalFk);
-                    }
+                    continue;
                 }
+
+                if (principalEntityType.BaseType != null)
+                {
+                    // #8973
+                    throw new InvalidOperationException(
+                        RelationalStrings.IncompatibleTableDerivedPrincipal(
+                            dependEntityType.Relational().TableName,
+                            dependEntityType.DisplayName(),
+                            principalEntityType.DisplayName(),
+                            principalEntityType.RootType().DisplayName()));
+                }
+                return true;
             }
 
             return false;

--- a/src/EFCore.Relational/Metadata/Internal/ConstraintNamer.cs
+++ b/src/EFCore.Relational/Metadata/Internal/ConstraintNamer.cs
@@ -80,7 +80,7 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Internal
         /// </summary>
         public static string GetDefaultName([NotNull] IProperty property)
         {
-            var sharedTablePrincipalPrimaryKeyProperty = property.FindSharedTablePrincipalPrimaryKeyProperty();
+            var sharedTablePrincipalPrimaryKeyProperty = property.FindSharedTableRootPrimaryKeyProperty();
             if (sharedTablePrincipalPrimaryKeyProperty != null)
             {
                 return sharedTablePrincipalPrimaryKeyProperty.Relational().ColumnName;

--- a/src/EFCore.Relational/Metadata/Internal/PropertyExtensions.cs
+++ b/src/EFCore.Relational/Metadata/Internal/PropertyExtensions.cs
@@ -15,10 +15,21 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Internal
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        public static IProperty FindSharedTablePrincipalPrimaryKeyProperty([NotNull] this IProperty property)
+        public static IProperty FindSharedTableRootPrimaryKeyProperty([NotNull] this IProperty property)
         {
-            var linkingRelationship = property.FindSharedTableLink();
-            return linkingRelationship?.PrincipalKey.Properties[linkingRelationship.Properties.IndexOf(property)];
+            var principalProperty = property;
+            while (true)
+            {
+                var linkingRelationship = principalProperty.FindSharedTableLink();
+                if (linkingRelationship == null)
+                {
+                    break;
+                }
+
+                principalProperty = linkingRelationship.PrincipalKey.Properties[linkingRelationship.Properties.IndexOf(principalProperty)];
+            }
+
+            return principalProperty == property ? null : principalProperty;
         }
 
         /// <summary>

--- a/src/EFCore.Relational/Properties/RelationalStrings.Designer.cs
+++ b/src/EFCore.Relational/Properties/RelationalStrings.Designer.cs
@@ -461,12 +461,12 @@ namespace Microsoft.EntityFrameworkCore.Internal
                 table, entityType, otherEntityType, keyName, primaryKey, otherName, otherPrimaryKey);
 
         /// <summary>
-        ///     Cannot use table '{table}' for entity type '{entityType}' since it is being used for entity type '{otherEntityType}' and there is no relationship between the primary key {primaryKey} and the primary key {otherPrimaryKey}.
+        ///     Cannot use table '{table}' for entity type '{entityType}' since it is being used for entity type '{otherEntityType}' and there is no relationship between their primary keys.
         /// </summary>
-        public static string IncompatibleTableNoRelationship([CanBeNull] object table, [CanBeNull] object entityType, [CanBeNull] object otherEntityType, [CanBeNull] object primaryKey, [CanBeNull] object otherPrimaryKey)
+        public static string IncompatibleTableNoRelationship([CanBeNull] object table, [CanBeNull] object entityType, [CanBeNull] object otherEntityType)
             => string.Format(
-                GetString("IncompatibleTableNoRelationship", nameof(table), nameof(entityType), nameof(otherEntityType), nameof(primaryKey), nameof(otherPrimaryKey)),
-                table, entityType, otherEntityType, primaryKey, otherPrimaryKey);
+                GetString("IncompatibleTableNoRelationship", nameof(table), nameof(entityType), nameof(otherEntityType)),
+                table, entityType, otherEntityType);
 
         /// <summary>
         ///     Cannot use table '{table}' for entity type '{dependentType}' since it has a relationship to a derived entity type '{principalType}'. Either point the relationship to the base type '{rootType}' or map '{dependentType}' to a different table.

--- a/src/EFCore.Relational/Properties/RelationalStrings.resx
+++ b/src/EFCore.Relational/Properties/RelationalStrings.resx
@@ -282,7 +282,7 @@
     <value>Cannot use table '{table}' for entity type '{entityType}' since it is being used for entity type '{otherEntityType}' and the name '{keyName}' of the primary key {primaryKey} does not match the name '{otherName}' of the primary key {otherPrimaryKey}.</value>
   </data>
   <data name="IncompatibleTableNoRelationship" xml:space="preserve">
-    <value>Cannot use table '{table}' for entity type '{entityType}' since it is being used for entity type '{otherEntityType}' and there is no relationship between the primary key {primaryKey} and the primary key {otherPrimaryKey}.</value>
+    <value>Cannot use table '{table}' for entity type '{entityType}' since it is being used for entity type '{otherEntityType}' and there is no relationship between their primary keys.</value>
   </data>
   <data name="IncompatibleTableDerivedPrincipal" xml:space="preserve">
     <value>Cannot use table '{table}' for entity type '{dependentType}' since it has a relationship to a derived entity type '{principalType}'. Either point the relationship to the base type '{rootType}' or map '{dependentType}' to a different table.</value>

--- a/src/EFCore.Specification.Tests/DataAnnotationTestBase.cs
+++ b/src/EFCore.Specification.Tests/DataAnnotationTestBase.cs
@@ -1032,6 +1032,7 @@ namespace Microsoft.EntityFrameworkCore
 
         protected class Login4
         {
+            public int Id { get; set; }
             public int Login4Id { get; set; }
             public string UserName { get; set; }
 
@@ -1042,6 +1043,7 @@ namespace Microsoft.EntityFrameworkCore
 
         protected class Profile4
         {
+            public int Id { get; set; }
             public int Profile4Id { get; set; }
             public string Name { get; set; }
             public string Email { get; set; }
@@ -1193,7 +1195,8 @@ namespace Microsoft.EntityFrameworkCore
 
         protected class Login9
         {
-            public int Login9Id { get; set; }
+            public int Id { get; set; }
+            public int? Login9Id { get; set; }
             public string UserName { get; set; }
 
             [ForeignKey("Login9Id")]
@@ -1202,7 +1205,8 @@ namespace Microsoft.EntityFrameworkCore
 
         protected class Profile9
         {
-            public int Profile9Id { get; set; }
+            public int Id { get; set; }
+            public int? Profile9Id { get; set; }
             public string Name { get; set; }
             public string Email { get; set; }
 
@@ -1219,15 +1223,17 @@ namespace Microsoft.EntityFrameworkCore
 
             Validate(modelBuilder);
 
-            Assert.True(GetProperty<Login10>(modelBuilder, "Id").IsForeignKey());
-            Assert.True(GetProperty<Profile10>(modelBuilder, "Id").IsForeignKey());
+            Assert.True(GetProperty<Login10>(modelBuilder, "FkId").IsForeignKey());
+            Assert.True(GetProperty<Profile10>(modelBuilder, "FkId").IsForeignKey());
         }
 
         public class Login10
         {
             public int Id { get; set; }
 
-            [ForeignKey("Id")]
+            public int FkId { get; set; }
+
+            [ForeignKey("FkId")]
             public virtual Profile10 Login { get; set; }
         }
 
@@ -1235,7 +1241,9 @@ namespace Microsoft.EntityFrameworkCore
         {
             public int Id { get; set; }
 
-            [ForeignKey("Id")]
+            public int FkId { get; set; }
+
+            [ForeignKey("FkId")]
             public Login10 User { get; set; }
         }
 

--- a/src/EFCore.SqlServer/Metadata/SqlServerPropertyAnnotations.cs
+++ b/src/EFCore.SqlServer/Metadata/SqlServerPropertyAnnotations.cs
@@ -42,8 +42,8 @@ namespace Microsoft.EntityFrameworkCore.Metadata
         /// </summary>
         public virtual string HiLoSequenceName
         {
-            get { return (string)Annotations.Metadata[SqlServerAnnotationNames.HiLoSequenceName]; }
-            [param: CanBeNull] set { SetHiLoSequenceName(value); }
+            get => (string)Annotations.Metadata[SqlServerAnnotationNames.HiLoSequenceName];
+            [param: CanBeNull] set => SetHiLoSequenceName(value);
         }
 
         /// <summary>
@@ -62,8 +62,8 @@ namespace Microsoft.EntityFrameworkCore.Metadata
         /// </summary>
         public virtual string HiLoSequenceSchema
         {
-            get { return (string)Annotations.Metadata[SqlServerAnnotationNames.HiLoSequenceSchema]; }
-            [param: CanBeNull] set { SetHiLoSequenceSchema(value); }
+            get => (string)Annotations.Metadata[SqlServerAnnotationNames.HiLoSequenceSchema];
+            [param: CanBeNull] set => SetHiLoSequenceSchema(value);
         }
 
         /// <summary>
@@ -142,7 +142,7 @@ namespace Microsoft.EntityFrameworkCore.Metadata
 
             if (Property.ValueGenerated != ValueGenerated.OnAdd)
             {
-                var sharedTablePrincipalPrimaryKeyProperty = Property.FindSharedTablePrincipalPrimaryKeyProperty();
+                var sharedTablePrincipalPrimaryKeyProperty = Property.FindSharedTableRootPrimaryKeyProperty();
                 if (sharedTablePrincipalPrimaryKeyProperty != null
                     && sharedTablePrincipalPrimaryKeyProperty.SqlServer().ValueGenerationStrategy == SqlServerValueGenerationStrategy.IdentityColumn)
                 {

--- a/src/EFCore/Infrastructure/ModelValidator.cs
+++ b/src/EFCore/Infrastructure/ModelValidator.cs
@@ -45,6 +45,7 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
             ValidateNonNullPrimaryKeys(model);
             ValidateNoShadowKeys(model);
             ValidateNoMutableKeys(model);
+            ValidateNoCycles(model);
             ValidateClrInheritance(model);
             ValidateChangeTrackingStrategy(model);
             ValidateDefiningNavigations(model);
@@ -151,6 +152,49 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
                 }
             }
         }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        protected virtual void ValidateNoCycles([NotNull] IModel model)
+        {
+            var unvalidatedEntityTypes = new HashSet<IEntityType>(model.GetEntityTypes());
+            foreach (var entityType in model.GetEntityTypes())
+            {
+                var primaryKey = entityType.FindPrimaryKey();
+                if (primaryKey != null)
+                {
+                    var identifyingForeignKeys = new Queue<IForeignKey>(
+                        entityType.FindForeignKeys(primaryKey.Properties).Where(fk => fk.PrincipalEntityType != entityType));
+                    while (identifyingForeignKeys.Count > 0)
+                    {
+                        var fk = identifyingForeignKeys.Dequeue();
+                        if (!fk.PrincipalKey.IsPrimaryKey()
+                            || !unvalidatedEntityTypes.Contains(fk.PrincipalEntityType))
+                        {
+                            continue;
+                        }
+
+                        if (fk.PrincipalEntityType == entityType)
+                        {
+                            throw new InvalidOperationException(CoreStrings.IdentifyingRelationshipCycle(entityType.DisplayName()));
+                        }
+
+                        foreach (var principalFk in fk.PrincipalEntityType.FindForeignKeys(fk.PrincipalKey.Properties))
+                        {
+                            if (principalFk.PrincipalEntityType != principalFk.DeclaringEntityType)
+                            {
+                                identifyingForeignKeys.Enqueue(principalFk);
+                            }
+                        }
+                    }
+                }
+
+                unvalidatedEntityTypes.Remove(entityType);
+            }
+        }
+
 
         /// <summary>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used

--- a/src/EFCore/Properties/CoreStrings.Designer.cs
+++ b/src/EFCore/Properties/CoreStrings.Designer.cs
@@ -2288,6 +2288,14 @@ namespace Microsoft.EntityFrameworkCore.Internal
                 GetString("IncompatiblePrincipalEntrySensitive", nameof(foreignKeyValues), nameof(dependentEntityType), nameof(keyValue), nameof(foundPrincipalEntityType), nameof(principalEntityType)),
                 foreignKeyValues, dependentEntityType, keyValue, foundPrincipalEntityType, principalEntityType);
 
+        /// <summary>
+        ///     The entity type '{entityType}' is part of a relationship cycle involving its primary key.
+        /// </summary>
+        public static string IdentifyingRelationshipCycle([CanBeNull] object entityType)
+            => string.Format(
+                GetString("IdentifyingRelationshipCycle", nameof(entityType)),
+                entityType);
+
         private static string GetString(string name, params string[] formatterNames)
         {
             var value = _resourceManager.GetString(name);

--- a/src/EFCore/Properties/CoreStrings.resx
+++ b/src/EFCore/Properties/CoreStrings.resx
@@ -966,4 +966,7 @@
   <data name="IncompatiblePrincipalEntrySensitive" xml:space="preserve">
     <value>The foreign key '{foreignKeyValues}' set on '{dependentEntityType}' with the key value '{keyValue}' matches an entity of type '{foundPrincipalEntityType}', however the principal entity type should be assignable to '{principalEntityType}'.</value>
   </data>
+  <data name="IdentifyingRelationshipCycle" xml:space="preserve">
+    <value>The entity type '{entityType}' is part of a relationship cycle involving its primary key.</value>
+  </data>
 </root>

--- a/test/EFCore.Relational.Tests/RelationalModelValidatorTest.cs
+++ b/test/EFCore.Relational.Tests/RelationalModelValidatorTest.cs
@@ -84,12 +84,12 @@ namespace Microsoft.EntityFrameworkCore
 
             VerifyError(
                 RelationalStrings.IncompatibleTableNoRelationship(
-                    "Schema.Table", entityB.DisplayName(), entityA.DisplayName(), "{'Id'}", "{'Id'}"),
+                    "Schema.Table", entityB.DisplayName(), entityA.DisplayName()),
                 model);
         }
 
         [Fact]
-        public virtual void Does_not_detect_duplicate_table_names_in_different_schema()
+        public virtual void Passes_for_duplicate_table_names_in_different_schema()
         {
             var model = new Model();
             var entityA = model.AddEntityType(typeof(A));
@@ -105,7 +105,7 @@ namespace Microsoft.EntityFrameworkCore
         }
 
         [Fact]
-        public virtual void Does_not_detect_duplicate_table_names_for_inherited_entities()
+        public virtual void Passes_for_duplicate_table_names_for_inherited_entities()
         {
             var model = new Model();
             var entityA = model.AddEntityType(typeof(A));
@@ -128,7 +128,7 @@ namespace Microsoft.EntityFrameworkCore
 
             VerifyError(
                 RelationalStrings.IncompatibleTableKeyNameMismatch(
-                    "Table", nameof(A), nameof(B), "Key", "{'Id'}", "PK_Table", "{'Id'}"),
+                    "Table", nameof(B), nameof(A), "PK_Table", "{'Id'}", "Key", "{'Id'}"),
                 modelBuilder.Model);
         }
 
@@ -164,6 +164,22 @@ namespace Microsoft.EntityFrameworkCore
             VerifyError(
                 RelationalStrings.DuplicateColumnNameDataTypeMismatch(
                     nameof(A), nameof(A.P0), nameof(B), nameof(B.P0), nameof(B.P0), "Table", "someInt", "default_int_mapping"), modelBuilder.Model);
+        }
+
+        [Fact]
+        public virtual void Detects_multiple_shared_table_roots()
+        {
+            var modelBuilder = CreateConventionalModelBuilder();
+
+            modelBuilder.Entity<A>().HasOne<B>().WithOne().IsRequired().HasForeignKey<A>(a => a.Id).HasPrincipalKey<B>(b => b.Id);
+            modelBuilder.Entity<A>().ToTable("Table");
+            modelBuilder.Entity<A>().HasOne<C>().WithOne().IsRequired().HasForeignKey<A>(a => a.Id).HasPrincipalKey<C>(b => b.Id);
+            modelBuilder.Entity<C>().HasBaseType((string)null).ToTable("Table");
+            modelBuilder.Entity<B>().ToTable("Table");
+
+            VerifyError(
+                RelationalStrings.IncompatibleTableNoRelationship("Table", nameof(C), nameof(B)),
+                modelBuilder.Model);
         }
 
         [Fact] // #8973
@@ -814,10 +830,6 @@ namespace Microsoft.EntityFrameworkCore
             entityType.Relational().DiscriminatorValue = entityType.Name;
         }
 
-        protected class C : A
-        {
-        }
-
         protected class Animal
         {
             public int Id { get; set; }
@@ -873,7 +885,7 @@ namespace Microsoft.EntityFrameworkCore
                         TestServiceFactory.Instance.Create<RelationalTypeMapperDependencies>(),
                         TestServiceFactory.Instance.Create<TestRelationalTypeMapper>())));
 
-        protected virtual ModelBuilder CreateConventionalModelBuilder()
+        protected override ModelBuilder CreateConventionalModelBuilder()
             => RelationalTestHelpers.Instance.CreateConventionBuilder();
     }
 }

--- a/test/EFCore.Sqlite.FunctionalTests/DataAnnotationSqliteTest.cs
+++ b/test/EFCore.Sqlite.FunctionalTests/DataAnnotationSqliteTest.cs
@@ -178,8 +178,9 @@ namespace Microsoft.EntityFrameworkCore
                 @"@p1='' (Nullable = false) (DbType = String)" + _eol,
                 Sql);
         }
-        
+
 #if !Test20
+
         public override void RequiredAttribute_for_property_throws_while_inserting_null_value()
         {
             base.RequiredAttribute_for_property_throws_while_inserting_null_value();

--- a/test/EFCore.Tests/Infrastructure/CoreModelValidatorTest.cs
+++ b/test/EFCore.Tests/Infrastructure/CoreModelValidatorTest.cs
@@ -146,6 +146,23 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
         }
 
         [Fact]
+        public virtual void Detects_relationship_cycle()
+        {
+            var modelBuilder = CreateConventionalModelBuilder();
+
+            modelBuilder.Entity<A>();
+            modelBuilder.Entity<B>();
+            modelBuilder.Entity<C>().HasBaseType((string)null);
+            modelBuilder.Entity<A>().HasOne<B>().WithOne().IsRequired().HasForeignKey<A>(a => a.Id).HasPrincipalKey<B>(b => b.Id);
+            modelBuilder.Entity<A>().HasOne<C>().WithOne().IsRequired().HasForeignKey<C>(a => a.Id).HasPrincipalKey<A>(b => b.Id);
+            modelBuilder.Entity<C>().HasOne<B>().WithOne().IsRequired().HasForeignKey<B>(a => a.Id).HasPrincipalKey<C>(b => b.Id);
+
+            VerifyError(
+                CoreStrings.IdentifyingRelationshipCycle(nameof(A)),
+                modelBuilder.Model);
+        }
+
+        [Fact]
         public virtual void Passes_on_escapable_foreign_key_cycles()
         {
             var model = new Model();

--- a/test/EFCore.Tests/Infrastructure/ModelValidatorTest.cs
+++ b/test/EFCore.Tests/Infrastructure/ModelValidatorTest.cs
@@ -76,6 +76,10 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
             public A AnotherA { get; set; }
         }
 
+        protected class C : A
+        {
+        }
+
         protected class D : A
         {
         }
@@ -175,5 +179,8 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
         }
 
         protected abstract IModelValidator CreateModelValidator();
+
+        protected virtual ModelBuilder CreateConventionalModelBuilder()
+            => InMemoryTestHelpers.Instance.CreateConventionBuilder();
     }
 }

